README
NAME
    DBIx::Class::Validation - Validate all data before submitting to your
    database.

SYNOPSIS
    In your base "DBIx::Class" package:

      __PACKAGE__->load_components(qw/... Validation/);

    And in your subclasses:

      __PACKAGE__->validation(
        module => 'FormValidator::Simple',
        profile => { ... },
        filters => 0,
        auto => 1,
      );

    And then somewhere else:

      eval{ $obj->validate() };
      if( my $results = $EVAL_ERROR ){
        ...
      }

METHODS
  validation
      __PACKAGE__->validation(
        module => 'FormValidator::Simple',
        profile => { ... },
        filters => 0,
        auto => 1,
      );

    Calls "validation_module", "validation_profile" and "validation_auto" if
    the corresponding argument is defined.

  validation_module
      __PACKAGE__->validation_module('Data::FormValidator');

    Sets the validation module to use. Any module that supports a check()
    method just like "Data::FormValidator"'s can be used here, such as
    "FormValidator::Simple".

    Defaults to FormValidator::Simple.

  validation_profile
      __PACKAGE__->validation_profile(
        { ... }
      );

    Sets the profile that will be passed to the validation module. Expects
    either a HASHREF or a reference to a subroutine. If it's a subref it
    will be passed the result row object as it's first parameter so that you
    can perform complex data validation for cases when you'd like to have
    access to the actual result.

    For example, you could use the following to return an error if the named
    field is not unique in the table:

        my $profile = sub {
            my $result = shift @_;
        
        return {
                required => [qw/email/],
                constraint_methods => {
                    email => sub {
                        my ($dvf, $val) = @_;
                        return $result->result_source->resultset->find({email=>$val}) ? 0:1;
                    },
                },
            };
        };

    Please note that the subref needs to return a hashref/arrayref suitable
    for use in the validation module you have chosen.

  validation_auto
      __PACKAGE__->validation_auto( 1 );

    Turns on and off auto-validation. This feature makes all UPDATEs and
    INSERTs call the "validate" method before doing anything.

    The default is for validation_auto is to be on.

  validation_filter
      __PACKAGE__->validation_filter( 1 );

    Turns on and off validation filters. When on, this feature will make all
    UPDATEs and INSERTs modify your data to that of the values returned by
    your validation modules "check" method. This is primarily meant for use
    with "Data::FormValidator" but may be used with any validation module
    that returns a results object that supports a "valid()" method just like
    "Data::FormValidator::Results".

    Filters modify your data, so use them carefully.

    The default is for validation_filter is to be off.

  validate
      $obj->validate();

    Validates all the data in the object against the pre-defined validation
    module and profile. If there is a problem then a hard error will be
    thrown. If you put the validation in an eval you can capture whatever
    the module's check() method returned.

EXTENDED METHODS
    The following "DBIx::Class::Row" methods are extended by this module:-

    insert
    update

SEE ALSO
    "DBIx::Class", "FormValidator::Simple", "Data::FormValidator"

AUTHOR
    Aran C. Deltac <bluefeet@cpan.org>

CONTRIBUTERS
    Tom Kirkpatrick <tkp@cpan.org> Christopher Laco <claco@cpan.org> John
    Napiorkowski <jjn1056@yahoo.com>

LICENSE
    You may distribute this code under the same terms as Perl itself.



Changes
Revision history for DBIx::Class::Validation

0.02004 Mon Jun 22 12:52:00 2009
    - Switch dependency to DateTime::Format::SQLite

0.02003 Mon Apr 22 21:41:23 2008
    - Fix test failure under SQL::Translator <= 0.07
    - Removed Build.PL now that Module::Install no longer supports it
    - Removed bogus tests entry in Makefile.PL

0.02002 Fri Mar 7 19:38:12 2008
    - Quick fix for CPANTS META/author

0.02001 Tue July 17 10:26:34 2007
    - Added build_requires for DateTime::Format::MySQL
    - Added WriteAll/DIST/PREOP for README
    - Fixed spelling test failure

0.02000 Sun June 24 20:04:11 2007
    - Converted to Class::Accessor::Grouped after mk_classdata
        snafu in DBIC 0.08
    - Validate now croaks instead of throw_exception to play
        well with DBIx::Class::Exception

0.01004 Sat May 12 10:45:23 2007
    - Use column accessors to set columns instead of set_column
        (John Napiorkowski)

0.01003 Tue May 08 18:49:11 2007
    - Fixed DBIC::Test comment snafu
    - Profiles can now be a subref that will dynamically generate the
        profile (John Napiorkowski)

0.01002 Sat May 05 11:38:16 2007
    - Converted to Module::Install
    - RT#25356 (bokutin): validation now sets filter
    - Converted tests away from deprecated compose_connection
    - Jockey cargo tests, TEST_AUTHOR, DBICTest
    - Much improved test coverage
    - Build/test no longer requires SQL::Translator

0.01001 Tue Sep 05 17:54:13 2006
    - FormValidator::Simple 0.17 is now required to fix profile destruction
        on first use
    - Removed Storable::dclone code since it's no longer needed

0.01000
    - initial release


t/lib/DBIC/Test.pm
# $Id: /mirror/trunk/DBIx-Class-Validation/t/lib/DBIC/Test.pm 4286 2008-04-23T01:15:37.302737Z claco  $
package DBIC::Test;
use strict;
use warnings;

BEGIN {
    # little trick by Ovid to pretend to subclass+exporter Test::More
    use base qw/Test::Builder::Module Class::Accessor::Grouped/;
    use Test::More;
    use File::Spec::Functions qw/catfile catdir/;

    @DBIC::Test::EXPORT = @Test::More::EXPORT;

    __PACKAGE__->mk_group_accessors('inherited', qw/db_dir db_file/);
};

__PACKAGE__->db_dir(catdir('t', 'var'));
__PACKAGE__->db_file('test.db');

## cribbed and modified from DBICTest in DBIx::Class tests
sub init_schema {
    my ($self, %args) = @_;
    my $db_dir  = $args{'db_dir'}  || $self->db_dir;
    my $db_file = $args{'db_file'} || $self->db_file;
    my $namespace = $args{'namespace'} || 'DBIC::TestSchema';
    my $db = catfile($db_dir, $db_file);

    eval 'use DBD::SQLite';
    if ($@) {
       BAIL_OUT('DBD::SQLite not installed');

        return;
    };

    eval 'use DBIC::Test::Schema';
    if ($@) {
        BAIL_OUT("Could not load DBIC::Test::Schema: $@");

        return;
    };

    unlink($db) if -e $db;
    unlink($db . '-journal') if -e $db . '-journal';
    mkdir($db_dir) unless -d $db_dir;

    my $dsn = 'dbi:SQLite:' . $db;
    my $schema = DBIC::Test::Schema->compose_namespace($namespace)->connect($dsn, undef, undef, {AutoCommit => 1});
    $schema->storage->on_connect_do([
        'PRAGMA synchronous = OFF',
        'PRAGMA temp_store = MEMORY'
    ]);

    __PACKAGE__->deploy_schema($schema, %args);
    __PACKAGE__->populate_schema($schema, %args) unless $args{'no_populate'};

    return $schema;
};

sub deploy_schema {
    my ($self, $schema, %options) = @_;
    my $eval = $options{'eval_deploy'};

    open IN, catfile('t', 'sql', 'test.sqlite.sql');
    my $sql;
    { local $/ = undef; $sql = <IN>; }
    close IN;
    eval {
        ($schema->storage->dbh->do($_) || print "Error on SQL: $_\n") for split(/;\n/, $sql);
    };
    if ($@ && !$eval) {
        die $@;
    };
};

sub clear_schema {
    my ($self, $schema, %options) = @_;

    foreach my $source ($schema->sources) {
        $schema->resultset($source)->delete_all;
    };
};

sub populate_schema {
    my ($self, $schema, %options) = @_;
    
    if ($options{'clear'}) {
        $self->clear_schema($schema, %options);
    };
};

1;


t/lib/Validator.pm
package Validator;
use strict;
use warnings;

sub new {
    return bless {}, shift;
};

sub check {
    
};

1;


t/lib/ValidationTest.pm
package # hide from PAUSE 
    ValidationTest;

use strict;
use warnings;
use ValidationTest::Schema;

sub init_schema {
    my $self = shift;
    my $db_file = "t/var/ValidationTest.db";

    unlink($db_file) if -e $db_file;
    unlink($db_file . "-journal") if -e $db_file . "-journal";
    mkdir("t/var") unless -d "t/var";

    my $dsn = "dbi:SQLite:${db_file}";
    
    my $schema = ValidationTest::Schema->compose_connection('ValidationTest' => $dsn);
    $schema->deploy();
    
    return $schema;
}

1;


lib/DBIx/Class/Validation.pm
# $Id: /mirror/trunk/DBIx-Class-Validation/lib/DBIx/Class/Validation.pm 6760 2009-06-22T12:52:54.917872Z ribasushi  $
package DBIx::Class::Validation;
use strict;
use warnings;
our $VERSION = '0.02004';

BEGIN {
    use base qw/DBIx::Class Class::Accessor::Grouped/;
    use English qw/-no_match_vars/;
    use FormValidator::Simple 0.17;
    use Carp qw/croak/;

    __PACKAGE__->mk_group_accessors('inherited', qw/
        validation_profile
        validation_auto
        validation_filter
        _validation_module_accessor
    /);
};
__PACKAGE__->validation_auto(1);
__PACKAGE__->validation_module('FormValidator::Simple');

=head1 NAME

DBIx::Class::Validation - Validate all data before submitting to your database.

=head1 SYNOPSIS

In your base L<"DBIx::Class"> package:

  __PACKAGE__->load_components(qw/... Validation/);

And in your subclasses:

  __PACKAGE__->validation(
    module => 'FormValidator::Simple',
    profile => { ... },
    filters => 0,
    auto => 1,
  );

And then somewhere else:

  eval{ $obj->validate() };
  if( my $results = $EVAL_ERROR ){
    ...
  }

=head1 METHODS

=head2 validation

  __PACKAGE__->validation(
    module => 'FormValidator::Simple',
    profile => { ... },
    filters => 0,
    auto => 1,
  );

Calls L</"validation_module">, L</"validation_profile"> and L</"validation_auto">
if the corresponding argument is defined.

=cut

sub validation {
    my ($self, %args) = @_;

    $self->validation_module($args{module}) if exists $args{module};
    $self->validation_profile($args{profile}) if exists $args{profile};
    $self->validation_auto($args{auto}) if exists $args{auto};
    $self->validation_filter($args{filter}) if exists $args{filter};
};

=head2 validation_module

  __PACKAGE__->validation_module('Data::FormValidator');

Sets the validation module to use.  Any module that supports a check() method
just like L<"Data::FormValidator">'s can be used here, such as
L<"FormValidator::Simple">.

Defaults to FormValidator::Simple.

=cut

sub validation_module {
    my ($self, $class) = @_;

    if ($class) {
        if (!eval "require $class") {
            $self->throw_exception("Unable to load the validation module '$class' because  $@");
        } elsif (!$class->can('check')) {
            $self->throw_exception("The '$class' module does not support the check() method");
        } else {
            $self->_validation_module_accessor($class->new);
        };
    };

    return ref $self->_validation_module_accessor;
};

=head2 validation_profile

  __PACKAGE__->validation_profile(
    { ... }
  );

Sets the profile that will be passed to the validation module.  Expects either
a HASHREF or a reference to a subroutine.  If it's a subref it will be passed
the result row object as it's first parameter so that you can perform complex
data validation for cases when you'd like to have access to the actual result.

For example, you could use the following to return an error if the named field
is not unique in the table:

    my $profile = sub {
        my $result = shift @_;
    
        return {
            required => [qw/email/],
            constraint_methods => {
                email => sub {
                    my ($dvf, $val) = @_;
                    return $result->result_source->resultset->find({email=>$val}) ? 0:1;
                },
            },
        };
    };

Please note that the subref needs to return a hashref/arrayref suitable for use
in the validation module you have chosen.

=head2 validation_auto

  __PACKAGE__->validation_auto( 1 );

Turns on and off auto-validation.  This feature makes all UPDATEs and
INSERTs call the L</"validate"> method before doing anything.

The default is for validation_auto is to be on.

=head2 validation_filter

  __PACKAGE__->validation_filter( 1 );

Turns on and off validation filters. When on, this feature will make all
UPDATEs and INSERTs modify your data to that of the values returned by
your validation modules C<check> method. This is primarily meant for use
with L<"Data::FormValidator"> but may be used with any validation module
that returns a results object that supports a C<valid()> method just
like L<"Data::FormValidator::Results">.

B<Filters modify your data, so use them carefully>.

The default is for validation_filter is to be off.

=head2 validate

  $obj->validate();

Validates all the data in the object against the pre-defined validation
module and profile.  If there is a problem then a hard error will be
thrown.  If you put the validation in an eval you can capture whatever
the module's check() method returned.

=cut

sub validate {
    my $self = shift;
    my %data = $self->get_columns;
    my $module = $self->validation_module;
    my $profile = $self->validation_profile;

    if (ref $profile eq 'CODE') {
        $profile = $profile->($self);
    };
    my $result = $module->check( \%data => $profile );

    if ($result->success) {
        if ($self->validation_filter && $result->can('valid')) {
            $self->$_($result->valid($_)) for ($result->valid);
        };
        return $result;
    } else {
        croak $result;
    };
};

=head1 EXTENDED METHODS

The following L<"DBIx::Class::Row"> methods are extended by this module:-

=over 4

=item insert

=cut

sub insert {
    my $self = shift;
    $self->validate if $self->validation_auto;
    $self->next::method(@_);
}

=item update

=cut

sub update {
    my $self = shift;
    $self->validate if $self->validation_auto;
    $self->next::method(@_);
};

1;
__END__

=back

=head1 SEE ALSO

L<"DBIx::Class">, L<"FormValidator::Simple">, L<"Data::FormValidator">

=head1 AUTHOR

Aran C. Deltac <bluefeet@cpan.org>

=head1 CONTRIBUTERS

Tom Kirkpatrick <tkp@cpan.org>
Christopher Laco <claco@cpan.org>
John Napiorkowski <jjn1056@yahoo.com>

=head1 LICENSE

You may distribute this code under the same terms as Perl itself.
